src/coreclr/interpreter/compiler.cpp

```diff
@@ -4612,37 +4911,643 @@ void InterpCompiler::EmitCall(CORINFO_RESOLVED_TOKEN* pConstrainedToken, bool re
         break;
     }
 
-    m_pLastNewIns->SetDVar(dVar);
-    m_pLastNewIns->SetSVar(CALL_ARGS_SVAR);
+    m_pLastNewIns->SetDVar(dVar);
+    m_pLastNewIns->SetSVar(CALL_ARGS_SVAR);
+
+    m_pLastNewIns->flags |= INTERP_INST_FLAG_CALL;
+    m_pLastNewIns->info.pCallInfo = (InterpCallInfo*)AllocMemPool0(sizeof (InterpCallInfo));
+    m_pLastNewIns->info.pCallInfo->pCallArgs = callArgs;
+
+    if (injectRet)
+    {
+        // Jmp to PInvoke was converted to normal pinvoke, so we need to inject a ret after the call
+        InterpType retType = GetInterpType(m_methodInfo->args.retType);
+        switch (retType)
+        {
+            case InterpTypeVT:
+            {
+                AddIns(INTOP_RET_VT);
+                m_pLastNewIns->SetSVar(dVar);
+                m_pLastNewIns->data[0] = m_pVars[dVar].size;
+                break;
+            }
+            case InterpTypeVoid:
+                AddIns(INTOP_RET_VOID);
+                break;
+            default:
+                AddIns(GetRetForType(retType));
+                m_pLastNewIns->SetSVar(dVar);
+                break;
+        }
+    }
+
+    if (callInfo.sig.isAsyncCall() && !tailcall && m_methodInfo->args.isAsyncCall()) // Async2 functions may need to suspend
+    {
+        EmitSuspend(callInfo, continuationContextHandling, pBB);
+    }
+}
+
+void SetSlotToTrue(TArray<bool, MemPoolAllocator> &gcRefMap, int32_t slotIndex)
+{
+    while (gcRefMap.GetSize() <= slotIndex)
+    {
+        int32_t growSize = gcRefMap.GetSize();
+        if (growSize < 16)
+            growSize = 16;
+        gcRefMap.GrowBy(growSize);
+    }
+    gcRefMap.Set(slotIndex, true);
+}
+
+void InterpCompiler::EmitSuspend(const CORINFO_CALL_INFO &callInfo, ContinuationContextHandling ContinuationContextHandling, InterpBasicBlock *pBB)
+{
+    CORINFO_LOOKUP_KIND kindForAllocationContinuation;
+    m_compHnd->getLocationOfThisType(m_methodHnd, &kindForAllocationContinuation);
+
+    bool needsKeepAlive = kindForAllocationContinuation.needsRuntimeLookup;
```

Comment:
JIT does not decide this condition based on `getLocationOfThisType`. Is it a bug in the JIT?
We use the various `CORINFO_GENERICS_CTXT_FROM_*` flags returned through `getMethodInfo` instead.